src/ipc/ipc4/handler.c

```diff
@@ -215,13 +215,14 @@ static bool is_any_ppl_active(void)
  *     ERROR Stop       EOS       |______\ SAVE
  *                                      /
  */
-int set_pipeline_state(struct ipc_comp_dev *ppl_icd, uint32_t cmd,
-		       bool *delayed)
+
+static int do_set_pipeline_state(struct ipc_comp_dev *ppl_icd, uint32_t cmd,
```

Comment:
@ranj063 ， this PR are trying to fix Seppo's SRC CI error（#7547）, params are not correctly propagated to SRC source buffer, I tried other ways, but not working。

Reply:
@ranj063, what I have seen with @singalsu's SRC support is
The capture looks like this SSP2 -> SRC -> host which is in topology represented with the following two pipelines:
pipeline0: ssp.copier
pipeline1: src -> host.copier

We have 2 pipelines and we batch start them, the call path is something like this:
ipc4_set_pipeline_state()
set_pipeline_state()
ipc4_pcm_params()
ipc4_pipeline_params()
 ipc4_comp_params()
 pipeline_prepare()

the ipc4_set_pipeline_state() calls the set_pipeline_state() in a for loop for each pipelines, starting from the sink side.
The ipc4_pipeline_params() is walking through the pipeline and sets the params, it does it by using `struct sof_ipc_pcm_params hw_params = {{ 0 }};` as initial params and this actually get set to the first buffer.
0x40001 : host.copier
0x80000 : SRC
0x40000 : ssp.copier
Just the relevant prints to no loose them. Modified `audio_stream_set_params()` to print the source and sink components

This is what happens:
```
[    3.334896] <wrn> ipc: ipc4_pcm_params: comp(1):1 0x40001 ipc4_pcm_params(): ENTER
[    3.334900] <wrn> ipc: ipc4_pipeline_params: comp(1):1 0x40001 ipc4_pipeline_params(): ENTER
[    3.334906] <wrn> ipc: audio_stream_set_params: comp(0):1 0x80000 audio_stream_set_params(): the SOURCE
[    3.334910] <wrn> ipc: audio_stream_set_params: comp(1):1 0x40001 audio_stream_set_params(): the SINK
[Heippa] audio_stream_set_params(): channels: 0 -> 0
[    3.334941] <wrn> ipc: audio_stream_set_params: comp(2):0 0x40000 audio_stream_set_params(): the SOURCE
[    3.334945] <wrn> ipc: audio_stream_set_params: comp(0):1 0x80000 audio_stream_set_params(): the SINK
[Heippa] audio_stream_set_params(): channels: 0 -> 2
[    3.334965] <wrn> ipc: ipc4_pipeline_params: comp(1):1 0x40001 ipc4_pipeline_params(): LEAVE
[    3.334968] <wrn> ipc: ipc4_pcm_params: comp(1):1 0x40001 ipc4_pcm_params(): After params
```
This was the ipc4_pipeline_params() run, notice that the buffer between host.copier and SRC is is receiving params = { 0 }
```
[    3.334985] <wrn> ipc: audio_stream_set_channels: comp(0):1 0x80000 audio_stream_set_channels(): the SOURCE
[    3.334990] <wrn> ipc: audio_stream_set_channels: comp(1):1 0x40001 audio_stream_set_channels(): the SINK
[Heippa] audio_stream_set_channels(): channels: 0 -> 2
[Heippa] audio_stream_set_params(): channels: 0 -> 2
[    3.335031] <inf> copier: copier_prepare: comp(1):1 0x40001 copier_prepare()
[    3.335041] <inf> src: src_prepare: comp(0):1 0x80000 src_prepare()
[    3.335046] <inf> src: src_params_general: comp(0):1 0x80000 src_params()
[    3.335051] <wrn> ipc: audio_stream_set_params: comp(0):1 0x80000 audio_stream_set_params(): the SOURCE
[    3.335055] <wrn> ipc: audio_stream_set_params: comp(1):1 0x40001 audio_stream_set_params(): the SINK
[Heippa] audio_stream_set_params(): channels: 2 -> 2
[    3.335068] <wrn> ipc: audio_stream_set_params: comp(2):0 0x40000 audio_stream_set_params(): the SOURCE
[    3.335073] <wrn> ipc: audio_stream_set_params: comp(0):1 0x80000 audio_stream_set_params(): the SINK
[Heippa] audio_stream_set_params(): channels: 2 -> 2
[    3.335085] <inf> src: src_params_general: comp(0):1 0x80000 src_params(), source_rate = 48000, sink_rate = 8000
[    3.335113] <wrn> ipc: ipc4_pcm_params: comp(1):1 0x40001 ipc4_pcm_params(): LEAVE
```
In the prepare part of the function (`pipeline_prepare()`) the host.copier config got corrected.
At this point the SRC is activated and would start processing, the pipeline0 however is yet to be configured:
```
[    3.335155] <wrn> ipc: ipc4_pcm_params: comp(2):0 0x40000 ipc4_pcm_params(): ENTER
[    3.335158] <wrn> ipc: ipc4_pipeline_params: comp(2):0 0x40000 ipc4_pipeline_params(): ENTER
[    3.335165] <wrn> ipc: audio_stream_set_params: comp(2):0 0x40000 audio_stream_set_params(): the SOURCE
[    3.335168] <wrn> ipc: audio_stream_set_params: comp(0):1 0x80000 audio_stream_set_params(): the SINK
[Heippa] audio_stream_set_params(): channels: 2 -> 0
[    3.335188] <wrn> ipc: ipc4_pipeline_params: comp(2):0 0x40000 ipc4_pipeline_params(): LEAVE

...

[    3.335265] <wrn> src: src_get_copy_limits: comp(2):0 0x40000 src_get_copy_limits(): the SOURCE
[    3.335270] <wrn> src: src_get_copy_limits: comp(0):1 0x80000 src_get_copy_limits(): the SINK
[crash check] src_get_copy_limits(): channels: 0

^^^ divide by zero crash ^^^
...

[    3.335300] <wrn> ipc: ipc4_pcm_params: comp(2):0 0x40000 ipc4_pcm_params(): After params
[    3.335316] <wrn> ipc: audio_stream_set_channels: comp(2):0 0x40000 audio_stream_set_channels(): the SOURCE
[    3.335321] <wrn> ipc: audio_stream_set_channels: comp(0):1 0x80000 audio_stream_set_channels(): the SINK
[Heippa] audio_stream_set_channels(): channels: 0 -> 2
[Heippa] audio_stream_set_params(): channels: 0 -> 2

...

[    3.335430] <inf> copier: copier_prepare: comp(2):0 0x40000 copier_prepare()
[    3.335460] <wrn> ipc: ipc4_pcm_params: comp(2):0 0x40000 ipc4_pcm_params(): LEAVE
```
At the first hunk the ipc4_pipeline_params() resets the buffer paams to 0 between dai.copier and SRC as this is the first (and only one) in the pipeline. It does it by clearing the already set config - as down in this path we have `comp_verify_params()` which always forces the params to be set and to me it only does that, not verifying anything.

Then the src's process is called (it is already active) and it will encounter a buffer on it's input with all params to be 0, causing a panic.

If I handle this then the firmware run further and the the prepare path will configure the params finally and consequent src prepare will not have issue.

I have suggested to @singalsu to silently ignore a buffer params setting in case all parameters are to be set to 0 as that is by itself is not valid.
Probably that is the correct fix, but we do have a race in this case and it is matter of timing that the component at the start of the pipeline will cause a crash or not.

Reply:
@ujfalusi good stuff, when doing cleanup can you add more inline comments which can include warnings to help other understand what can/can't be done in each phase and why.

Reply:
@ujfalusi thanks for teh detailed explanation. But isnt a simpler fix just to modify the params before this call to copier_update_params() https://github.com/thesofproject/sof/blob/b81f1309e843ece412fb5c8852962af2cf807a90/src/audio/copier/copier.c#L621

to read the params from the basecfg so that we dont overwrite the buffer with 0 params?

Reply:
@ujfalusi there's no params to be passed with IPC4. EVery module has it in its basecfg no?

Reply:
@ranj063, are you suggesting that `ipc4_pipeline_params()` path should be NOP and dropped?
At the end we have two issues brought to surface by the SRC (big thanks to @singalsu for it!):
1. the params of the buffer on the sink side of the first component in a pipeline got reset to 0 in the `ipc4_pipeline_params()`
1.1. this is 'corrected' at the `pipeline_prepare()` path
1.2. this got caught by the SRC module which actually uses and looks at the source buffer's params

2. when triggering (starting) the pipelines got configured (`ipc4_pipeline_params() and `pipeline_prepare()`) _and_ set to run (`pipeline_trigger()`) individually and as a batch process
2.2. this can lead to the following race condition:
pipeline1 configured, and set to active: components start their processing
pipeline0 half way configured
component from pipeline1 is executing it's process function and try to access/interact with the pipeline0, but that is not yet have been configured, not even set to run
This is only problematic in components at the junction of the pipelines and only for those which needs to know the config of the other side (like SRC).

The issue is not that the pipeline0 is not active, that is fine, but it is not even configured fully and that is a problem.

For issue 1 we could ignore params set which would like to zero the configuration (invalid by any means), or get the initial params settings from basecfg of the head of the pipeline (I guess) - which might not be a copier, or remove the `ipc4_pipeline_params()` path altogether (you said that it is not needed in IPC4), or re-architect it completely to do things differently.

For issue 2, this PR provides a solution.

Reply:
What this PR does is:
configure and prepare all pipelines first
trigger them

In this way when the trigger is run on the pipelines, all of the pipelines are fully configured, prepared.

Reply:
@ujfalusi I think this PR is good to do even if we were to drop pipeline params. Let's go with it.

Reply:
Isn't it one of the cases that should be fixed by all pipeline actions carried out in the pipeline task context? That was the mechanism that protected us from half-configured pipelines being triggered?

Reply:
replying to myself: not quite, with IPC3 we also processed pipeline parameters in the IPC task context, the difference was only, that there setting pipeline parameters were a separate IPC message.

Reply:
@lyakh, the pipeline is configured (pipeline 1 in the example) and it is triggered, but the configuration is not yet finished on pipeline 0 and pipeline 0 is not yet triggered (because it should be configured first).

In a simplified view, we have two pipelines connected: pipeline0, pipeline1 and we do atm:
configure pipeline1
trigger pipeline1
configure pipeline0
trigger pipeline0

with this PR:
configure pipeline1
configure pipeline0
trigger pipeline1
trigger pipeline0

Reply:
There is one thing which might need to be considered however, I think: if we have a really long pipeline then the trigger round will be 'delayed' to the point when all pipelines have been configured already while currently the first pipeline is triggered right after the first configuration.
The last trigger will happen at the same time, but the time distribution will differ.

If this is a problem than probably have some sort of one step later trigger
```
for (i = 0; i < num_pipelines; i++) {
	configure_pipeline(i);
	if (i)
		trigger_pipeline(i - 1);
}

trigger_pipeline(i - 1); /* trigger_pipeline(num_pipelines - 1); */
```
so we run the trigger with a delay after the configure -> connected pipes are configured